refactored signal decorator with async

ColinGrasley opened the pull request.

grugna committed:
Merge pull request #15 from chicagopcdc/pcdc_dev

Pcdc dev

grugna committed:
Merge pull request #17 from chicagopcdc/pcdc_dev

Pcdc dev

grugna committed:
Merge pull request #21 from chicagopcdc/pcdc_dev

Pcdc dev

ColinGrasley committed:
alternative to signal

ColinGrasley committed:
change to using multiprocessing so that spawned threads are killed

ColinGrasley committed:
readded required libraries

ColinGrasley committed:
Refactored timeout decorator with asyncio

modification is required to update the repos that use FenceClientManager with await and async.

grugna commented:
@ColinGrasley another way to handle both async and sync without having to change the other repo is [this](https://github.com/chicagopcdc/pcdcutils/blob/9acdde4a359bce9256843270230ec8fad3493605/pcdcutils/gen3/__init__.py#L87-L112)
Could you try a similar implementation?

ColinGrasley committed:
added wrapper to avoid unnecessary modification for calls to FenceClientManager methods 

changed it so you do not have to add async/await to calls of the FenceClientManager class

ColinGrasley commented:
It may need more testing with credentials, I would need someone to generate credentials for me in order for me to test it more thoroughly

grugna commented:
Use this: https://uc-cri.slack.com/archives/D08RKAXUELX/p1748891018745429

kellcuc commented on pcdcutils/client/__init__.py:
```diff
@@ -1,36 +1,66 @@
 
-        @functools.wraps(func)
-        def wrapper(*args, **kwargs):
-            signal.signal(signal.SIGALRM, _handle_timeout)
```
Removed this, but i could not find it used anywhere else in chicagopcdc

kellcuc commented on pcdcutils/client/__init__.py:
```diff
@@ -1,36 +1,66 @@
 
 from gen3.auth import Gen3Auth, Gen3AuthError
 
+import asyncio
```
good use of async for future compatibility

kellcuc commented on pcdcutils/client/__init__.py:
```diff
@@ -1,36 +1,66 @@
 import errno
 import os
-import signal
```
Consistent throughout removal of signal, not used throughout codebase.

kellcuc commented on pcdcutils/client/__init__.py:
```diff
@@ -1,36 +1,66 @@
-
 def timeout(seconds=10, error_message=os.strerror(errno.ETIME)):
     def decorator(func):
-        def _handle_timeout(signum, frame):
```
Won't break, and call has also been removed

kellcuc commented on pcdcutils/client/__init__.py:
```diff
@@ -1,36 +1,66 @@
+def syncify_async(async_func):
+    @functools.wraps(async_func)
+    def wrapper(*args, **kwargs):
+        import threading
```
Was there a reason for `import threading` to be inside of `syncify_async`? Unless there is a specific reason, i personally like following Black standards and having them grouped.

ColinGrasley replied:
oops, yeah I see your point, I should have moved that up top

kellcuc commented:
Good work.

Just a small thing, it’s usually best to keep all imports at the top of the file (unless a specific reason), so it’s easy to see everything in one place. https://github.com/chicagopcdc/pcdcutils/pull/26#discussion_r2191219015

Otherwise, everything looks great!

ColinGrasley committed:
moved import

kellcuc commented:
pcdc loads and i can navigate without issue. 

Last few suggestions:
1. You may want to change origin of intern_ to pcdc_dev, so that you have the most current code. Also, it helps to avoid conflict when merging.
2. May want to add a test file, I wrote one, which i will share below, but due to item 1, you will need to first pull the most current pcdc_dev for pcdcutils.
3. May want to update the poetry.lock or pyproject.toml file, but does not look to be needed
4. May want to add comments in the source code throughout for other developers. I try to not make my comments not very complex, and fairly straightforward, but that is me :)

For the test_utils.py
1. Create a file called `fence_keys.env` and store it in `/tests/keys/`, add the below madeup values to the file.
```
FAKE_FENCE_URL=http://fence-service
FAKE_CLIENT_ID=w2R8mk4YE7PQfKGq5uh1NXdjTLoz3Ic6e9VUaOiZ
FAKE_CLIENT_SECRET=vn1zS8WyXMLOHp0ARjETe6ZKkV4g3fYNq9BIsUuJwhCbmVDlF2QoG7itJr5PsaTh
```
2. add the below to the top of the `/tests/test_utils.py` file
```
from pcdcutils.client import FenceClientManager
import time

def load_fence_env(file_path):
    env_vars = {}
    with open(file_path) as f:
        for line in f:
            if "=" in line:
                key, value = line.strip().split("=", 1)
                env_vars[key] = value
    return env_vars


# Keys in this file are fake
env_key_path = os.getcwd() + "/tests/keys/fence_keys.env"
env = load_fence_env(env_key_path)

FAKE_FENCE_URL = env["FAKE_FENCE_URL"]
FAKE_CLIENT_ID = env["FAKE_CLIENT_ID"]
FAKE_CLIENT_SECRET = env["FAKE_CLIENT_SECRET"]
```

Add this after the last test in the file

```
def test_fence_client_manager_success(monkeypatch):
    # We create a fake Gen3Auth class to avoid making real network requests.
    class FakeGen3Auth:
        def __init__(self, endpoint, client_credentials, client_scopes):
            # Save the inputs so we can check them if needed
            self.endpoint = endpoint
            self.client_credentials = client_credentials
            self.client_scopes = client_scopes
            # We set a fake access token value
            self._access_token = "fake-token"

        def get_access_token(self):
            # This returns the fake token instead of making a real API call
            return self._access_token

    # Replace (monkeypatch) the real Gen3Auth with our fake one inside FenceClientManager
    monkeypatch.setattr("pcdcutils.client.Gen3Auth", FakeGen3Auth)

    # Create a FenceClientManager instance using our test config values
    client = FenceClientManager(
        base_url=FAKE_FENCE_URL,
        client_id=FAKE_CLIENT_ID,
        client_secret=FAKE_CLIENT_SECRET,
    )

    # Run authenticate() — this should use our FakeGen3Auth and set the auth object
    client.authenticate()

    # Check that the client now says it's authenticated
    assert client.is_authenticated()

    # Get a token from our fake auth — should match our fake token value
    token = client.get_auth_token()
    assert token == "fake-token"

def test_fence_client_manager_invalid(monkeypatch):
    # Create a client without any credentials to simulate a misconfigured state
    client = FenceClientManager(base_url=None, client_id=None, client_secret=None)

    # Run authenticate() — since credentials are missing, this should not set auth
    client.authenticate()

    # Check that the client is not authenticated
    assert not client.is_authenticated()

    # Get token — because there's no auth, we expect an empty string (not an error)
    token = client.get_auth_token()
    assert token == ""

def test_fence_client_manager_timeout(monkeypatch):
    # We create a fake Gen3Auth that simulates a very slow initialization
    class SlowGen3Auth:
        def __init__(self, endpoint, client_credentials, client_scopes):
            # Sleep for 5 seconds to simulate slow network or processing
            time.sleep(5)
            self._access_token = "slow-token"

        def get_access_token(self):
            return self._access_token

    # Replace the real Gen3Auth with our slow fake version
    monkeypatch.setattr("pcdcutils.client.Gen3Auth", SlowGen3Auth)

    # Create a client using valid config
    client = FenceClientManager(
        base_url=FAKE_FENCE_URL,
        client_id=FAKE_CLIENT_ID,
        client_secret=FAKE_CLIENT_SECRET,
    )

    # Call authenticate() — it should fail due to hitting the timeout limit
    # We expect it to raise an exception (e.g., TimeoutError or similar)
    with pytest.raises(Exception):
        client.authenticate()
```

Feel free to edit as you like, this just an idea.

I wrote steps on how to run these tests - https://github.com/chicagopcdc/pcdcutils/blob/pcdc_dev/README.md

ColinGrasley committed:
Merge branch 'pcdc_dev' into intern_

ColinGrasley commented:
@kellcuc While doing testing with using Asyncio i figured out that Asyncio will not actually stop the running of Gen3Auth if it is stuck in an infinite loops of trying to connect. This is due to it not giving back control to the event loop, so a blocking call such as sleep will not be interrupted and stopped. While testing I found that using a time.sleep() call will keep the thread still active in the background until the sleep timer is finished. I could not find a way to actively kill it. I managed to reimplement the timeout using multiprocessing, where it actually has the ability to kill the running gen3Auth call.

ColinGrasley commented:
to my understanding multiprocessor has quite a bit of overhead as it creates a process, I have implemented it with multiprocessing and knows it works with the fakeGen3Auth class. should I use it with multiprocessor? Or go back to trying to find another way of implementing? I pushed to the intern_ branch so you can see what i have. I cannot mock the real authentication Gen3Auth function with a fake one to simulate because the multiprocessing library is not compatible with mocking.

ColinGrasley committed:
implementation using the multiprocess library, needs refining with tetsing, as monkey patch is not compatible with multiprocessing

ColinGrasley committed:
added explanation to testing file

kellcuc commented:
@ColinGrasley 
Yeah, that totally makes sense, great job figuring that out. You are right: `asyncio` can’t stop something like a time.sleep() or a blocking network call because it’s not giving control back to the event loop. 

Using multiprocessing to force-kill it does work, it comes with some tradeoffs. It’s heavier, can be harder to debug, and like you said, mocking doesn’t really play nice with it.

In catching rare edge cases, like if Gen3Auth hangs completely, I’d probably lean toward keeping it simple with asyncio and threads. We can’t kill the function mid-run, but we can at least detect when it takes too long and log it or handle it gracefully. Having logging would be helpful here, or a counter. In the counter method, when x is hit, the call would close, and we could run one last non async call, which would kill on fail. 

Nice work digging into this.
